File: nodejs/math/point.js
const angle = require("./angle");
const { clamp } = require("./utils");

class Point {
    x; y;
    static ORIGIN = new Point(0, 0);

    static fromAngle(angle, magnitude = 1) {
        let x = magnitude * -1 * Math.sin(angle);
        let y = magnitude * Math.cos(angle);
        return new Point(x, y);
    }

    static average(points) {
        let x = 0;
        let y = 0;
        let c = points.length;
        for (let i = 0; i < c; i++) {
            x += points[i].x;
            y += points[i].y;
        }
        return new Point(x/c, y/c);
    }

    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    dist(point) {
        let dx = point.x - this.x;
        let dy = point.y - this.y;
        return Math.sqrt(dx*dx + dy*dy);
    }

    get magnitude() { return this.dist(Point.ORIGIN); }

    get angle() {
        if (this.x == 0 && this.y == 0) return angle.radians(0);
        return angle.radians(Math.atan2(this.x * -1, this.y));
    }
}

module.exports = Point;

File: python/math_utils/point.py
import math
from math_utils.angle import radians
from math_utils.utils import clamp

class Point():
    @classmethod
    def fromAngle(cls, angle, magnitude=1):
        x = magnitude * -1 * math.sin(angle)
        y = magnitude * math.cos(angle)
        return cls(x, y)

    @classmethod
    def average(cls, points):
        x = 0
        y = 0
        for p in points:
            x = x + p.x
            y = y + p.y
        c = len(points)
        return cls(x/c, y/c)

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def dist(self, point):
        dx = point.x - self.x
        dy = point.y - self.y
        return math.sqrt(dx*dx + dy*dy)

    def getMagnitude(self):
        return self.dist(Point.ORIGIN)

    def getAngle(self):
        if self.x == 0 and self.y == 0:
            return radians(0)
        return radians(math.atan2(self.x * -1, self.y))

    def keepInsideRectangle(self, rect):
        x = clamp(self.x, rect.origin.x, rect.corner.x)
        y = clamp(self.y, rect.origin.y, rect.corner.y)
        return Point(x, y)

Point.ORIGIN = Point(0, 0)
